src/storage-client/src/controller.rs

```diff
@@ -886,25 +890,57 @@ where
                 .await
                 .unwrap();
 
-            let (write, mut read) = persist_client
-                .open(metadata.data_shard)
+            let write = persist_client
+                .open_writer(metadata.data_shard)
                 .await
                 .expect("invalid persist usage");
 
-            // Advance the collection's `since` as requested.
-            if let Some(since) = &description.since {
-                read.downgrade_since(since).await;
-            }
+            // Construct the handle in a separate block to ensure all error paths are diverging
+            let since_handle = {
+                let mut handle = persist_client
+                    .open_critical_since(
+                        metadata.data_shard,
+                        PersistClient::CONTROLLER_CRITICAL_SINCE,
+                    )
+                    .await
+                    .expect("invalid persist usage");
+
+                let since = description
+                    .since
+                    .clone()
+                    .unwrap_or_else(|| handle.since().clone());
+
+                // We should only continue if we can fence out any other processes
+                let our_epoch = self.state.envd_epoch;
+                if handle.opaque() <= &our_epoch {
+                    let cur = *handle.opaque();
+                    let result = handle
+                        .maybe_compare_and_downgrade_since(&cur, (&our_epoch, &since))
+                        .await;
+                    match result {
+                        Some(Ok(_)) => {
+                            assert_eq!(handle.opaque(), &our_epoch);
+                            handle
+                        }
+                        Some(Err(epoch)) => {
+                            mz_ore::halt!("fenced by envd @ {epoch}. ours = {our_epoch}")
```

Comment:
ah, looks like there might be some race condition here, where the epoch is not whatever we got at registration time, but also is still < our_epoch. I guess just toss this whole bit in a loop?

Reply:
The halt in the CI happens further down which is why it's so weird. It just has an identical message but you can see the module path The halt happening there means that envd successfully went past this section, which takes ownership of a since handle, but then someone flipped the epoch back to zero!

Reply:
That sounds like something is dropping the stash tables and recreating them. Another test running in parallel?